### DISCUSSION TITLE
What kind of Object will be passed to the InstrumentationContext returned by beginFieldExecution/beginFieldCompletion/beginFieldListCompletion?

### DISCUSSION THREAD

**Discussion by xenomachina** [graphql-java/graphql-java#3703] [Q&A] [answered]
We're migrating from graphql-java 21.5 to 22.1.

The Instrumentation interface has changed quite a bit, and I'm having some trouble figuring out the correct way to migrate some of the methods. I looked at https://github.com/graphql-java/graphql-java/issues/3609, which primarily talks about `beginFieldFetching`. I didn't have significant trouble migrating that method, but there are a few others I'm unsure about.

The three methods I'm trying to figure out, along with what I think are their replacements, are:

- `beginField` → `beginFieldExecution`
- `beginFieldComplete` → `beginFieldCompletion`
- `beginFieldListComplete` → `beginFieldListCompletion`

In all three of these cases, aside from the name changing, the (only) other change was the return type switching from `InstrumentationContext<ExecutionResult>` to `InstrumentationContext<Object>`.

In my implementations of these methods I was using implementations of `InstrumentationContext<ExecutionResult>` that read the contents of the `ExecutionResult` passed into `onCompleted`.

Now that the new methods return an `InstrumentationContext<Object>`, I'm not sure what is going to be passed into `onCompleted`. If it isn't an `ExecutionResult`, what is it? I don't know if I missed it, but I couldn't find any explanation of what to expect from the Java doc of either `Instrumentation` or `InstrumentationContext`.

**Comment by bbakerman** [comment, marked as answer]
We made this breaking change for performance reasons.  We looked at the code and if we fetched some value (say a scalar like a String or an object or list say) we would wrap it into an `ExecutionResult` and pass it back into the engine.

The engine would then "unwrap" that `ExecutionResult` to get the data value.  We would do this for all values in all objects and lists.

Needless to say this was not efficient and in 22 we decided to bite the bullet and break the Instrumentation contract as well as other places in the engine.

So the object you get back is the object fetched for that field.  A scalar, or object or list.  This is the same object if before you did `graphql.ExecutionResult#getData`

> In my implementations of these methods I was using implementations of InstrumentationContext<ExecutionResult> that read the contents of the ExecutionResult passed into onCompleted.

So where you called `er.getData()` now just use the object - its the same thing

**Reply by xenomachina** [reply]
Thank you! That makes sense.
